### DISCUSSION TITLE
Are multi-target TCA apps still a thing?

### DISCUSSION THREAD

**Discussion by jberstler** [pointfreeco/swift-composable-architecture#3151] [Q&A]
I watched an enormous amount of the TCA video series going waaaay back to the very beginning. One of the ideas that really got me interested was that each feature or screen of my app could be its own module that compiles on its own, and can be tested in complete isolation. However, I have struggled to get this working in practice. I also notice that it isn't really discussed much any more in newer Pointfree content, and from what I've seen, isn't really demonstrated in the example apps.

So I think it is worth asking... is this still a thing?

It's entirely possible that what I'm running up against is a lack of understanding in how Xcode builds projects. In my head, external libraries (like TCA) should be added to a target's "Frameworks, Libraries, and Embedded Content" while dependencies on other targets defined in my project should go in Build Phases > Target Dependencies. Invariably this results in either undefined symbol compiler warnings when I try to integrate multiple targets, or having TCA types defined in multiple places and getting a runtime error about trying to resolve which one will get used.

While I'm sure I could provide an example project to demonstrate this, I'd rather start with a general discussion about it. Is this something people do in practice? Is there a guide on the proper way to integrate targets? Are there examples of TCA apps that are built this way?

Thanks in advance!

**Comment by mbrandonw** [comment]
Hi @jberstler, modularity of TCA features is definitely still a thing, and a huge selling point of TCA. We discuss modularity in general, without being specific to TCA, in these two episodes:

* https://www.pointfree.co/episodes/ep171-modularization-part-1
* https://www.pointfree.co/episodes/ep172-modularization-part-2

In particular, we use SPM for modularizing an app because it's simplest, but you can also do the same in Xcode with frameworks.

And those principles apply to TCA too. We even open sourced a [very large TCA app](https://github.com/pointfreeco/isowords) that is built in this modularized fashion.

If you have specific questions about modularizing then please do ask.

**Reply by jberstler** [reply]
Thanks for the pointers! I'll check them out.

**Reply by gcohen-dev** [reply]
Hi @jberstler, 
The way to initialize a package is a bit different now, you can check this article: https://medium.com/@guycohendev/local-spm-mastering-modularization-with-swift-package-manager-xcode-15-e37b14c36199

**Comment by Narsail** [comment]
@jberstler My project has shared Reducers between an iOS and watchOS target. Before modularising my project it were just files added to both targets, now they are part of a core framework. 

I haven't used a guide to arrive there but rather just 'did it'. But let me know if you other questions about it.

**Reply by ahbou** [reply]
Hello,
I'm setting up a similar project and I have trouble sharing data between the iOS app and the WatchOS one.
I'm using WCSession. 
Have you achieved this?

**Reply by Narsail** [reply]
Oh I never saw your message. Sorry! I'm using CloudKit to share data. I'm planning to dive into WCSession in the future, but haven't implemented it yet.

**Reply by ahbou** [reply]
I ended up using WCSession

**Comment by jberstler** [comment]
@mbrandonw While I still don't know what I was doing wrong with Framework targets, I was easily able to convert these all to local swift packages and everything builds and runs just as expected. Thanks!

**Comment by igniscyan** [comment]


**Reply by gcohen-dev** [reply]
Hi @igniscyan, 
The initial setup is simpler now, check out: 
https://medium.com/@guycohendev/local-spm-mastering-modularization-with-swift-package-manager-xcode-15-e37b14c36199
